File: infra/lambda-functions/words.py
import json
import logging
from db_utils import execute_query, lambda_response, extract_path_params, extract_query_params, extract_body

logger = logging.getLogger()
logger.setLevel(logging.INFO)

def handler(event, context):
    """Handle word-related API requests"""
    try:
        http_method = event['httpMethod']
        path_params = extract_path_params(event)
        query_params = extract_query_params(event)
        
        # Determine the route based on path
        resource_path = event.get('resource', '')
        
        if http_method == 'GET':
            if '/languages/{languageId}/words' in resource_path:
                return get_language_words(path_params.get('languageId'), query_params.get('userId'))
            elif '{wordId}' in resource_path:
                return get_word(path_params.get('wordId'), query_params.get('userId'))
            else:
                return get_words(query_params)
                
        elif http_method == 'POST':
            return create_word(extract_body(event))
            
        elif http_method == 'PUT':
            return update_word(path_params.get('wordId'), extract_body(event))
            
        elif http_method == 'DELETE':
            return delete_word(path_params.get('wordId'), query_params.get('userId'))
            
        else:
            return lambda_response(405, {'error': 'Method not allowed'})
            
    except Exception as e:
        logger.error(f"Error in words handler: {str(e)}")
        return lambda_response(500, {'error': str(e)})

def get_words(query_params: dict):
    """Get all words with optional filtering, including translations"""
    language_id = query_params.get('languageId')
    user_id = query_params.get('userId')
    
    if language_id and user_id:
        # Verify user owns the language
        verify_query = "SELECT id FROM app_8b514_languages WHERE id = %s AND user_id = %s"
        verify_result = execute_query(verify_query, (language_id, user_id))
        
        if not verify_result:
            return lambda_response(404, {'error': 'Language not found or access denied'})
        
        query = """
            SELECT 
                w.*,
                COALESCE(
                    json_agg(
                        json_build_object(
                            'id', t.id,
                            'language_code', t.language_code,
                            'meaning', t.meaning,
                            'created_at', t.created_at
                        ) ORDER BY t.created_at
                    ) FILTER (WHERE t.id IS NOT NULL),
                    '[]'::json
                ) as app_8b514_translations
            FROM app_8b514_words w
            LEFT JOIN app_8b514_translations t ON w.id = t.word_id
            WHERE w.language_id = %s
            GROUP BY w.id
            ORDER BY w.created_at DESC
        """
        result = execute_query(query, (language_id,))
    elif language_id:
        query = """
            SELECT 
                w.*,
                COALESCE(
                    json_agg(
                        json_build_object(
                            'id', t.id,
                            'language_code', t.language_code,
                            'meaning', t.meaning,
                            'created_at', t.created_at
                        ) ORDER BY t.created_at
                    ) FILTER (WHERE t.id IS NOT NULL),
                    '[]'::json
                ) as app_8b514_translations
            FROM app_8b514_words w
            LEFT JOIN app_8b514_translations t ON w.id = t.word_id
            WHERE w.language_id = %s
            GROUP BY w.id
            ORDER BY w.created_at DESC
        """
        result = execute_query(query, (language_id,))
    else:
        query = """
            SELECT 
                w.*,
                COALESCE(
                    json_agg(
                        json_build_object(
                            'id', t.id,
                            'language_code', t.language_code,
                            'meaning', t.meaning,
                            'created_at', t.created_at
                        ) ORDER BY t.created_at
                    ) FILTER (WHERE t.id IS NOT NULL),
                    '[]'::json
                ) as app_8b514_translations
            FROM app_8b514_words w
            LEFT JOIN app_8b514_translations t ON w.id = t.word_id
            GROUP BY w.id
            ORDER BY w.created_at DESC
        """
        result = execute_query(query)
    
    return lambda_response(200, result)

def get_language_words(language_id: str, user_id: str = None):
    """Get words for a specific language with translations"""
    if not language_id:
        return lambda_response(400, {'error': 'Language ID is required'})
    
    if user_id:
        # Verify user owns the language
        verify_query = "SELECT id FROM app_8b514_languages WHERE id = %s AND user_id = %s"
        verify_result = execute_query(verify_query, (language_id, user_id))
        
        if not verify_result:
            return lambda_response(404, {'error': 'Language not found or access denied'})
    
    # Get words with their translations
    query = """
        SELECT 
            w.*,
            COALESCE(
                json_agg(
                    json_build_object(
                        'id', t.id,
                        'language_code', t.language_code,
                        'meaning', t.meaning,
                        'created_at', t.created_at
                    ) ORDER BY t.created_at
                ) FILTER (WHERE t.id IS NOT NULL),
                '[]'::json
            ) as app_8b514_translations
        FROM app_8b514_words w
        LEFT JOIN app_8b514_translations t ON w.id = t.word_id
        WHERE w.language_id = %s
        GROUP BY w.id
        ORDER BY w.created_at DESC
    """
    result = execute_query(query, (language_id,))
    
    return lambda_response(200, result)

def get_word(word_id: str, user_id: str = None):
    """Get a specific word with translations"""
    if not word_id:
        return lambda_response(400, {'error': 'Word ID is required'})
    
    if user_id:
        # Verify user owns the language that contains this word
        query = """
            SELECT w.* FROM app_8b514_words w
            JOIN app_8b514_languages l ON w.language_id = l.id
            WHERE w.id = %s AND l.user_id = %s
        """
        result = execute_query(query, (word_id, user_id))
    else:
        query = "SELECT * FROM app_8b514_words WHERE id = %s"
        result = execute_query(query, (word_id,))
    
    if not result:
        return lambda_response(404, {'error': 'Word not found'})
    
    word = result[0]
    
    # Get translations for this word
    translations_query = "SELECT * FROM app_8b514_translations WHERE word_id = %s ORDER BY created_at"
    translations = execute_query(translations_query, (word_id,))
    word['translations'] = translations or []
    
    return lambda_response(200, word)

def create_word(word_data: dict):
    """Create a new word"""
    required_fields = ['language_id', 'word', 'ipa']
    
    for field in required_fields:
        if field not in word_data:
            return lambda_response(400, {'error': f'Missing required field: {field}'})
    
    # Verify user owns the language if user_id provided
    if 'user_id' in word_data:
        verify_query = "SELECT id FROM app_8b514_languages WHERE id = %s AND user_id = %s"
        verify_result = execute_query(verify_query, (word_data['language_id'], word_data['user_id']))
        
        if not verify_result:
            return lambda_response(404, {'error': 'Language not found or access denied'})
    
    # Set defaults
    pos = word_data.get('pos', [])
    is_root = word_data.get('is_root', False)
    embedding = word_data.get('embedding')
    
    query = """
        INSERT INTO app_8b514_words (language_id, word, ipa, pos, is_root, embedding)
        VALUES (%s, %s, %s, %s, %s, %s)
        RETURNING *
    """
    
    result = execute_query(
        query,
        (
            word_data['language_id'],
            word_data['word'],
            word_data['ipa'],
            pos,
            is_root,
            embedding
        )
    )
    
    word_id = result[0]['id']
    
    # Handle translations if provided
    if 'translations' in word_data:
        for translation in word_data['translations']:
            if translation.get('meaning'):  # Only insert non-empty translations
                insert_translation_query = """
                    INSERT INTO app_8b514_translations (word_id, language_code, meaning)
                    VALUES (%s, %s, %s)
                """
                execute_query(insert_translation_query, (
                    word_id,
                    translation.get('language_code', 'en'),
                    translation['meaning']
                ))
    
    return lambda_response(201, result[0])

def update_word(word_id: str, word_data: dict):
    """Update an existing word"""
    if not word_id:
        return lambda_response(400, {'error': 'Word ID is required'})
    
    # Verify user owns the language if user_id provided
    if 'user_id' in word_data:
        verify_query = """
            SELECT w.id FROM app_8b514_words w
            JOIN app_8b514_languages l ON w.language_id = l.id
            WHERE w.id = %s AND l.user_id = %s
        """
        verify_result = execute_query(verify_query, (word_id, word_data['user_id']))
        
        if not verify_result:
            return lambda_response(404, {'error': 'Word not found or access denied'})
    
    # Build dynamic update query
    update_fields = []
    params = []
    
    if 'word' in word_data:
        update_fields.append('word = %s')
        params.append(word_data['word'])
    
    if 'ipa' in word_data:
        update_fields.append('ipa = %s')
        params.append(word_data['ipa'])
    
    if 'pos' in word_data:
        update_fields.append('pos = %s')
        params.append(word_data['pos'])
    
    if 'is_root' in word_data:
        update_fields.append('is_root = %s')
        params.append(word_data['is_root'])
    
    if 'embedding' in word_data:
        update_fields.append('embedding = %s')
        params.append(word_data['embedding'])
    
    if not update_fields:
        return lambda_response(400, {'error': 'No fields to update'})
    
    params.append(word_id)
    query = f"""
        UPDATE app_8b514_words 
        SET {', '.join(update_fields)}
        WHERE id = %s
        RETURNING *
    """
    
    result = execute_query(query, tuple(params))
    
    if not result:
        return lambda_response(404, {'error': 'Word not found'})
    
    # Handle translations update if provided
    if 'translations' in word_data:
        # Delete existing translations
        delete_translations_query = "DELETE FROM app_8b514_translations WHERE word_id = %s"
        execute_query(delete_translations_query, (word_id,))
        
        # Insert new translations
        for translation in word_data['translations']:
            if translation.get('meaning'):  # Only insert non-empty translations
                insert_translation_query = """
                    INSERT INTO app_8b514_translations (word_id, language_code, meaning)
                    VALUES (%s, %s, %s)
                """
                execute_query(insert_translation_query, (
                    word_id,
                    translation.get('language_code', 'en'),
                    translation['meaning']
                ))
    
    return lambda_response(200, result[0])

def delete_word(word_id: str, user_id: str = None):
    """Delete a word"""
    if not word_id:
        return lambda_response(400, {'error': 'Word ID is required'})
    
    if user_id:
        # Verify user owns the language that contains this word
        query = """
            DELETE FROM app_8b514_words 
            WHERE id = %s AND language_id IN (
                SELECT id FROM app_8b514_languages WHERE user_id = %s
            )
            RETURNING id
        """
        result = execute_query(query, (word_id, user_id))
    else:
        query = "DELETE FROM app_8b514_words WHERE id = %s RETURNING id"
        result = execute_query(query, (word_id,))
    
    if not result:
        return lambda_response(404, {'error': 'Word not found or access denied'})
    
    return lambda_response(200, {'message': 'Word deleted successfully', 'id': result[0]['id']})

File: infra/lambda-functions-nodejs/src/words.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { executeQuery, lambdaResponse, extractPathParams, extractQueryParams, extractBody } from './db-utils';

export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
  try {
    const httpMethod = event.httpMethod;
    const pathParams = extractPathParams(event);
    const queryParams = extractQueryParams(event);
    const resourcePath = event.resource || '';

    switch (httpMethod) {
      case 'GET':
        if (resourcePath.includes('/languages/{languageId}/words')) {
          return await getLanguageWords(pathParams.languageId, queryParams.userId);
        } else if (pathParams.wordId) {
          return await getWord(pathParams.wordId, queryParams.userId);
        } else {
          return await getWords(queryParams);
        }
      case 'POST':
        return await createWord(extractBody(event));
      case 'PUT':
        return await updateWord(pathParams.wordId, extractBody(event));
      case 'DELETE':
        return await deleteWord(pathParams.wordId, queryParams.userId);
      default:
        return lambdaResponse(405, { error: 'Method not allowed' });
    }
  } catch (error) {
    console.error('Error in words handler:', error);
    return lambdaResponse(500, { error: (error as Error).message });
  }
};

async function getWords(queryParams: Record<string, string>): Promise<APIGatewayProxyResult> {
  const languageId = queryParams.languageId;
  const userId = queryParams.userId;

  let query: string;
  let params: any[] = [];

  const baseQuery = `
    SELECT 
      w.*,
      COALESCE(
        json_agg(
          json_build_object(
            'id', t.id,
            'language_code', t.language_code,
            'meaning', t.meaning,
            'created_at', t.created_at
          ) ORDER BY t.created_at
        ) FILTER (WHERE t.id IS NOT NULL),
        '[]'::json
      ) as app_8b514_translations
    FROM app_8b514_words w
    LEFT JOIN app_8b514_translations t ON w.id = t.word_id
  `;

  if (languageId && userId) {
    const verifyQuery = 'SELECT id FROM app_8b514_languages WHERE id = $1 AND user_id = $2';
    const verifyResult = await executeQuery(verifyQuery, [languageId, userId]);

    if (!verifyResult.length) {
      return lambdaResponse(404, { error: 'Language not found or access denied' });
    }

    query = `${baseQuery} WHERE w.language_id = $1 GROUP BY w.id ORDER BY w.created_at DESC`;
    params = [languageId];
  } else if (languageId) {
    query = `${baseQuery} WHERE w.language_id = $1 GROUP BY w.id ORDER BY w.created_at DESC`;
    params = [languageId];
  } else {
    query = `${baseQuery} GROUP BY w.id ORDER BY w.created_at DESC`;
  }

  const result = await executeQuery(query, params);
  return lambdaResponse(200, result);
}

async function getLanguageWords(languageId: string, userId?: string): Promise<APIGatewayProxyResult> {
  if (!languageId) {
    return lambdaResponse(400, { error: 'Language ID is required' });
  }

  if (userId) {
    const verifyQuery = 'SELECT id FROM app_8b514_languages WHERE id = $1 AND user_id = $2';
    const verifyResult = await executeQuery(verifyQuery, [languageId, userId]);

    if (!verifyResult.length) {
      return lambdaResponse(404, { error: 'Language not found or access denied' });
    }
  }

  const query = `
    SELECT 
      w.*,
      COALESCE(
        json_agg(
          json_build_object(
            'id', t.id,
            'language_code', t.language_code,
            'meaning', t.meaning,
            'created_at', t.created_at
          ) ORDER BY t.created_at
        ) FILTER (WHERE t.id IS NOT NULL),
        '[]'::json
      ) as app_8b514_translations
    FROM app_8b514_words w
    LEFT JOIN app_8b514_translations t ON w.id = t.word_id
    WHERE w.language_id = $1
    GROUP BY w.id
    ORDER BY w.created_at DESC
  `;

  const result = await executeQuery(query, [languageId]);
  return lambdaResponse(200, result);
}

async function getWord(wordId: string, userId?: string): Promise<APIGatewayProxyResult> {
  if (!wordId) {
    return lambdaResponse(400, { error: 'Word ID is required' });
  }

  let query: string;
  let params: any[];

  if (userId) {
    query = `
      SELECT w.* FROM app_8b514_words w
      JOIN app_8b514_languages l ON w.language_id = l.id
      WHERE w.id = $1 AND l.user_id = $2
    `;
    params = [wordId, userId];
  } else {
    query = 'SELECT * FROM app_8b514_words WHERE id = $1';
    params = [wordId];
  }

  const result = await executeQuery(query, params);

  if (!result.length) {
    return lambdaResponse(404, { error: 'Word not found' });
  }

  const word = result[0];

  const translationsQuery = 'SELECT * FROM app_8b514_translations WHERE word_id = $1 ORDER BY created_at';
  const translations = await executeQuery(translationsQuery, [wordId]);
  word.translations = translations || [];

  return lambdaResponse(200, word);
}

async function createWord(wordData: any): Promise<APIGatewayProxyResult> {
  const requiredFields = ['language_id', 'word', 'ipa'];

  for (const field of requiredFields) {
    if (!wordData[field]) {
      return lambdaResponse(400, { error: `Missing required field: ${field}` });
    }
  }

  if (wordData.user_id) {
    const verifyQuery = 'SELECT id FROM app_8b514_languages WHERE id = $1 AND user_id = $2';
    const verifyResult = await executeQuery(verifyQuery, [wordData.language_id, wordData.user_id]);

    if (!verifyResult.length) {
      return lambdaResponse(404, { error: 'Language not found or access denied' });
    }
  }

  const pos = wordData.pos || [];
  const isRoot = wordData.is_root || false;
  const embedding = wordData.embedding || null;

  const query = `
    INSERT INTO app_8b514_words (language_id, word, ipa, pos, is_root, embedding)
    VALUES ($1, $2, $3, $4, $5, $6)
    RETURNING *
  `;

  const result = await executeQuery(query, [
    wordData.language_id,
    wordData.word,
    wordData.ipa,
    pos,
    isRoot,
    embedding
  ]);

  const wordId = result[0].id;
  const translations: any[] = [];

  if (wordData.translations) {
    for (const translation of wordData.translations) {
      if (translation.meaning) {
        const insertTranslationQuery = `
          INSERT INTO app_8b514_translations (word_id, language_code, meaning)
          VALUES ($1, $2, $3)
          RETURNING *
        `;
        const translationResult = await executeQuery(insertTranslationQuery, [
          wordId,
          translation.language_code || 'en',
          translation.meaning
        ]);
        if (translationResult.length) {
          translations.push(translationResult[0]);
        }
      }
    }
  }

  const wordResult = result[0];
  wordResult.translations = translations;

  return lambdaResponse(201, wordResult);
}

async function updateWord(wordId: string, wordData: any): Promise<APIGatewayProxyResult> {
  if (!wordId) {
    return lambdaResponse(400, { error: 'Word ID is required' });
  }

  if (wordData.user_id) {
    const verifyQuery = `
      SELECT w.id FROM app_8b514_words w
      JOIN app_8b514_languages l ON w.language_id = l.id
      WHERE w.id = $1 AND l.user_id = $2
    `;
    const verifyResult = await executeQuery(verifyQuery, [wordId, wordData.user_id]);

    if (!verifyResult.length) {
      return lambdaResponse(404, { error: 'Word not found or access denied' });
    }
  }

  const updateFields: string[] = [];
  const params: any[] = [];
  let paramIndex = 1;

  if (wordData.word) {
    updateFields.push(`word = $${paramIndex++}`);
    params.push(wordData.word);
  }

  if (wordData.ipa) {
    updateFields.push(`ipa = $${paramIndex++}`);
    params.push(wordData.ipa);
  }

  if (wordData.pos) {
    updateFields.push(`pos = $${paramIndex++}`);
    params.push(wordData.pos);
  }

  if (wordData.is_root !== undefined) {
    updateFields.push(`is_root = $${paramIndex++}`);
    params.push(wordData.is_root);
  }

  if (wordData.embedding !== undefined) {
    updateFields.push(`embedding = $${paramIndex++}`);
    params.push(wordData.embedding);
  }

  if (!updateFields.length) {
    return lambdaResponse(400, { error: 'No fields to update' });
  }

  params.push(wordId);
  const query = `
    UPDATE app_8b514_words 
    SET ${updateFields.join(', ')}
    WHERE id = $${paramIndex}
    RETURNING *
  `;

  const result = await executeQuery(query, params);

  if (!result.length) {
    return lambdaResponse(404, { error: 'Word not found' });
  }

  if (wordData.translations) {
    await executeQuery('DELETE FROM app_8b514_translations WHERE word_id = $1', [wordId]);

    for (const translation of wordData.translations) {
      if (translation.meaning) {
        const insertTranslationQuery = `
          INSERT INTO app_8b514_translations (word_id, language_code, meaning)
          VALUES ($1, $2, $3)
        `;
        await executeQuery(insertTranslationQuery, [
          wordId,
          translation.language_code || 'en',
          translation.meaning
        ]);
      }
    }
  }

  return lambdaResponse(200, result[0]);
}

async function deleteWord(wordId: string, userId?: string): Promise<APIGatewayProxyResult> {
  if (!wordId) {
    return lambdaResponse(400, { error: 'Word ID is required' });
  }

  let query: string;
  let params: any[];

  if (userId) {
    query = `
      DELETE FROM app_8b514_words 
      WHERE id = $1 AND language_id IN (
        SELECT id FROM app_8b514_languages WHERE user_id = $2
      )
      RETURNING id
    `;
    params = [wordId, userId];
  } else {
    query = 'DELETE FROM app_8b514_words WHERE id = $1 RETURNING id';
    params = [wordId];
  }

  const result = await executeQuery(query, params);

  if (!result.length) {
    return lambdaResponse(404, { error: 'Word not found or access denied' });
  }

  return lambdaResponse(200, { message: 'Word deleted successfully', id: result[0].id });
}